src/R2N.jl

```diff
@@ -0,0 +1,412 @@
+- `subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = ShiftedLBFGSSolver`: the subsolver to solve the shifted system. Default is `JSOSolvers.ShiftedLBFGSSolver` which is the exact solver.
+- `subsolver_verbose::Int = 0`: if > 0, display iteration information every `subsolver_verbose` iteration of the subsolver if CG is selected.
+
+# Output
+The value returned is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Callback
+The callback is called at each iteration.
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.x`: current iterate;
+- `solver.gx`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.dual_feas`: norm of current gradient;
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
+
+# Examples
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = R2N(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = R2NSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+
+mutable struct R2NSolver{
+  T,
+  V,
+  Op <: AbstractLinearOperator{T},
+  Op2 <: AbstractLinearOperator{T},
+  Sub <: Union{KrylovSolver{T, T, V}, ShiftedLBFGSSolver},
+} <: AbstractOptimizationSolver
+  x::V
+  cx::V
+  gx::V
+  gn::V
+  σ::T
+  μ::T
+  H::Op
+  opI::Op2
+  Hs::V
+  s::V
+  obj_vec::V # used for non-monotone behaviour
+  subsolver_type::Sub
+  cgtol::T
+end
+
+function R2NSolver(
+  nlp::AbstractNLPModel{T, V};
+  non_mono_size = 1,
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+) where {T, V}
+  nvar = nlp.meta.nvar
+  x = V(undef, nvar) # vs similar(nlp.meta.x0)
+  cx = V(undef, nvar)
+  gx = V(undef, nvar)
+  gn = isa(nlp, QuasiNewtonModel) ? V(undef, nvar) : V(undef, 0)
+  Hs = V(undef, nvar)
+  H = isa(nlp, QuasiNewtonModel) ? nlp.op : hess_op!(nlp, x, Hs)
+  opI = opEye(T, nvar)
+  Op = typeof(H)
+  Op2 = typeof(opI)
+  σ = zero(T)
+  μ = zero(T)
+  s = V(undef, nvar)
+  cgtol = one(T) # must be ≤ 1.0
+  obj_vec = fill(typemin(T), non_mono_size)
+  subsolver =
```

Comment:
How do you check that the model must be an LBFGSModel is order to use the ShiftedLBFGSSolver?

There should be a unit test that checks the logic.

Reply:
I used this, 
```julia
  if (solver.subsolver_type isa ShiftedLBFGSSolver && !isa(nlp, LBFGSModel))
    error("Unsupported subsolver type, ShiftedLBFGSSolver is only can be used by LBFGSModel")
```
is this good, I agree, I will add a test for it in unittest